command/server.go

```diff
@@ -286,6 +286,22 @@ func (c *ServerCommand) Run(args []string) int {
 		}
 	}
 
+	// If the backend supports service discovery, run service discovery
+	if coreConfig.HAPhysical != nil {
+		sd, ok := coreConfig.HAPhysical.(physical.ServiceDiscovery)
+		if ok {
+			if err := sd.UpdateAdvertiseAddr(coreConfig.AdvertiseAddr); err != nil {
+				c.Ui.Error(fmt.Sprintf("Error configuring service discovery: %v", err))
+				return 1
+			}
+
+			if err := sd.RunServiceDiscovery(c.ShutdownCh); err != nil {
```

Comment:
I think the functionality of UpdateAdvertiseAddr could be folded into RunServiceDiscovery -- `RunServiceDiscovery(c.ShutdownCh, coreConfig.AdvertiseAddr)`